### PR TITLE
Slack notification for unaccredited provider recruitment

### PR DESCRIPTION
**Pull request by JR-G**
## Context

We have a number of providers who are now closed. We want to receive a notification if they continue to recruit candidates on Apply.

## Changes proposed in this pull request

Introduce a service that will be called as part of `MakeOffer`. This will check to see if the provider is closed and if so, will make a slack post alerting us they are currently recruiting.

### CODE REVIEW

File: app/services/notify_of_unaccredited_provider_recruitment.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
+class NotifyOfUnaccreditedProviderRecruitment
+  UNACCREDITED_PROVIDERS = [
+    69,
```
Comment (duncanjbrown) [inline app/services/notify_of_unaccredited_provider_recruitment.rb:3]: Are these database IDs? Better to use provider codes I think, for portability?

File: app/services/notify_of_unaccredited_provider_recruitment.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
+class NotifyOfUnaccreditedProviderRecruitment
+  UNACCREDITED_PROVIDERS = [
```
Comment (duncanjbrown) [inline app/services/notify_of_unaccredited_provider_recruitment.rb:2]: `CLOSED_PROVIDERS`? Providers that aren't accredited can still recruit.

File: app/services/make_offer.rb [hunk 3]
```diff
@@ -38,6 +38,7 @@ def save!
         ).call!
 
         SendNewOfferEmailToCandidate.new(application_choice:).call
+        NotifyOfUnaccreditedProviderRecruitment.new(application_choice: application_choice).check_for_unaccredited_provider
```
Comment (duncanjbrown) [inline app/services/make_offer.rb:41]: `NotifyOfRecruitmentByClosedProviders`?

**Review by duncanjbrown — COMMENTED** [review]
Is the language of "accreditation" a bit confusing here because we do allow schools etc to recruit candidates via manage?

File: app/services/make_offer.rb (hunk 3 above)
Comment (duncanjbrown) [reply app/services/make_offer.rb:41]: Worth sticking to the calling convention on services which is `X.new(configuration_etc).call`

Comment (JR-G) [reply app/services/make_offer.rb]: 🧹

File: app/services/notify_of_offer_by_closed_providers.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    end
+  end
+
+private
+
+  def send_slack_notification
+    message = ":bangbang: #{provider.name} has made an offer to a candidate – This provider is currently closed."
+    url = Rails.application.routes.url_helpers.support_interface_application_form_url(application_choice.application_form)
+
+    SlackNotificationWorker.perform_async(message, url, 'bat_provider_changes')
```
Comment (duncanjbrown) [inline app/services/notify_of_offer_by_closed_providers.rb:88]: What's the third arg here?

Comment (JR-G) [reply app/services/notify_of_offer_by_closed_providers.rb:88]: it's the slack channel – otherwise defaults to `twd_apply_support`

Comment (duncanjbrown) [reply app/services/notify_of_offer_by_closed_providers.rb:88]: ha — sorry, you've changed below! carry on!

File: app/workers/slack_notification_worker.rb [hunk 5]
```diff
@@ -18,10 +18,10 @@ def hyperlink(text, url)
     "<#{url}|#{text}>"
   end
 
-  def post_to_slack(text)
+  def post_to_slack(text, channel)
     if HostingEnvironment.production?
       slack_message = text
-      slack_channel = '#twd_apply_support'
+      slack_channel = channel || '#twd_apply_support'
```
Comment (duncanjbrown) [inline app/workers/slack_notification_worker.rb:24]: inconsistency with `#` in the caller?

File: app/services/notify_of_offer_by_closed_providers.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  def call
+    if CLOSED_PROVIDERS.include?(provider.code)
+      send_slack_notification
+    end
+  end
+
+private
+
+  def send_slack_notification
+    message = ":bangbang: #{provider.name} has made an offer to a candidate – This provider is currently closed."
```
Comment (duncanjbrown) [inline app/services/notify_of_offer_by_closed_providers.rb:85]:
```suggestion
    message = ":bangbang: #{provider.name} has made an offer to a candidate – this provider is currently closed."
```

File: app/workers/slack_notification_worker.rb (hunk 5 above)
Comment (JR-G) [reply app/workers/slack_notification_worker.rb]: ✅

File: app/workers/slack_notification_worker.rb [hunk 7]
```diff
@@ -18,13 +18,19 @@ def hyperlink(text, url)
     "<#{url}|#{text}>"
   end
 
-  def post_to_slack(text)
+  def ensure_correct_channel_format(channel)
+    return if channel.nil?
+
+    channel.prepend('#') if channel.first != '#'
```
Comment (duncanjbrown) [inline app/workers/slack_notification_worker.rb:24]: 👌

**Review by duncanjbrown — APPROVED** [review]
LGTM!

**Pull request merged**